File: src/main/java/algo/backtrack/GenerateParenthesis_22.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.List;

public class GenerateParenthesis_22 {
    public static List<String> generateParenthesis(int n) {
        var result = new ArrayList<String>();
        backtrack(n, n, "", result);
        return result;
    }

    private static void backtrack(int leftCount, int rightCount, String paren, List<String> result) {
        if (leftCount == 0 && rightCount == 0) {
            result.add(paren);
            return;
        }
        if (leftCount > 0) {
            backtrack(leftCount - 1, rightCount, paren + "(", result);
        }
        if (rightCount > leftCount) {
            backtrack(leftCount, rightCount - 1, paren + ")", result);
        }
    }

    public static void main(String[] args) {
        System.out.println(generateParenthesis(1));
        System.out.println(generateParenthesis(2));
        System.out.println(generateParenthesis(3));
    }
}

File: src/main/kotlin/algo/backtrack/GenerateParenthesis_22.kt
package algo.backtrack

class KtGenerateParenthesis_22 {
    fun generateParenthesis(n: Int): List<String> {
        val paren: MutableList<String> = mutableListOf()
        fun backtrack(leftCount: Int, rightCount: Int, str: String) {
            if (leftCount == 0 && rightCount == 0) {
                paren.add(str)
                return
            }
            if (leftCount == 0) {
                if (rightCount > 0) {
                    backtrack(leftCount, rightCount - 1, "$str)")
                }
            } else {
                backtrack(leftCount - 1, rightCount, "$str(")
                // right paren should not exceed left paren
                if (rightCount > leftCount) {
                    backtrack(leftCount, rightCount - 1, "$str)")
                }
            }
        }
        backtrack(n, n, "")
        return paren
    }
}

fun main() {
    val gen = KtGenerateParenthesis_22()
    println(gen.generateParenthesis(1))
    println(gen.generateParenthesis(2))
    println(gen.generateParenthesis(3))
}